File: src/main/services/BepInExSetup.ts
import { app } from 'electron'
import fs from 'fs'
import path from 'path'
import { BEPINEX_FOLDER } from '../constants'
import { createTempDir, extractArchiveAsync, removeDirIfInside } from './ModArchiveExtractor'
import type { InstallProgress, Result } from '../../shared/types'


export function bepinexRoot(gameRoot: string): string {
  return path.join(gameRoot, BEPINEX_FOLDER)
}


export function detectBepInEx(gameRoot: string): boolean {
  
  const root = bepinexRoot(gameRoot)
  if (fs.existsSync(path.join(root, 'config', 'BepInEx.cfg'))) return true
  if (fs.existsSync(path.join(root, 'core', 'BepInEx.Core.dll'))) return true
  return fs.existsSync(path.join(root, 'BepInEx.dll'))
}


export function bepinexArchivePath(): string | null {
  const candidates: string[] = []
  if (app.isPackaged) {
    candidates.push(path.join(process.resourcesPath, 'BepInEx.zip'))
  } else {
    candidates.push(path.join(app.getAppPath(), 'resources', 'BepInEx.zip'))
  }
  for (const c of candidates) {
    if (fs.existsSync(c)) return c
  }
  return null
}


export async function installBepInEx(
  gameRoot: string,
  onProgress?: (p: InstallProgress) => void
): Promise<Result<boolean>> {
  if (process.platform === 'win32') {
    return installBepInExWin32(gameRoot, onProgress)
  }
  return {
    ok: false,
    error: 'Automatic BepInEx installation is not supported on this platform. Please drop the BepInEx folder into the game directory manually.'
  }
}


function emit(onProgress: ((p: InstallProgress) => void) | undefined, message: string, percent = 0): void {
  onProgress?.({ stage: 'setup-bepinex', percent, message })
}


function moveChildren(srcDir: string, dstDir: string): void {
  for (const name of fs.readdirSync(srcDir)) {
    const src = path.join(srcDir, name)
    const dst = path.join(dstDir, name)
    fs.mkdirSync(path.dirname(dst), { recursive: true })
    try {
      fs.renameSync(src, dst)
    } catch {
      fs.cpSync(src, dst, { recursive: true })
      fs.rmSync(src, { recursive: true, force: true })
    }
  }
}


function moveToGameRoot(tempRoot: string, gameRoot: string): void {
  const top = fs.readdirSync(tempRoot, { withFileTypes: true })
  if (top.length === 1 && top[0].isDirectory() && top[0].name !== BEPINEX_FOLDER) {
    moveChildren(path.join(tempRoot, top[0].name), gameRoot)
    return
  }
  moveChildren(tempRoot, gameRoot)
}


async function installBepInExWin32(
  gameRoot: string,
  onProgress?: (p: InstallProgress) => void
): Promise<Result<boolean>> {
  const archive = bepinexArchivePath()
  if (!archive) {
    return { ok: false, error: 'BepInEx package is missing from the application bundle.' }
  }
  const tempRoot = createTempDir(gameRoot)
  try {
    emit(onProgress, 'Extracting BepInEx...', 10)
    await extractArchiveAsync(archive, tempRoot)
    emit(onProgress, 'Moving into game folder...', 60)
    moveToGameRoot(tempRoot, path.resolve(gameRoot))
    emit(onProgress, 'BepInEx installed.', 100)
    return { ok: true, value: detectBepInEx(gameRoot) }
  } catch (err) {
    return { ok: false, error: err instanceof Error ? err.message : String(err) }
  } finally {
    removeDirIfInside(gameRoot, tempRoot)
  }
}

File: RBBPMM.Wpf/RBBPMM.Core/BepInExSetup.cs
namespace RBBPMM.Core;

/// <summary>BepInEx detection + install (7z). Ported from src/main/services/BepInExSetup.ts.</summary>
public static class BepInExSetup
{
    /// <summary>Override the bundled BepInEx.zip location (defaults to the app base dir).</summary>
    public static string? BepInExArchivePathOverride { get; set; }

    public static string BepinexRoot(string gameRoot) => Path.Combine(gameRoot, Constants.BepInExFolder);

    public static bool DetectBepInEx(string gameRoot)
    {
        var root = BepinexRoot(gameRoot);
        if (File.Exists(Path.Combine(root, "config", "BepInEx.cfg"))) return true;
        if (File.Exists(Path.Combine(root, "core", "BepInEx.Core.dll"))) return true;
        return File.Exists(Path.Combine(root, "BepInEx.dll"));
    }

    public static string? ResolveBepInExArchive()
    {
        if (!string.IsNullOrEmpty(BepInExArchivePathOverride) && File.Exists(BepInExArchivePathOverride))
            return BepInExArchivePathOverride;
        string[] candidates =
        [
            Path.Combine(AppContext.BaseDirectory, "BepInEx.zip"),
            Path.Combine(AppContext.BaseDirectory, "resources", "BepInEx.zip"),
        ];
        foreach (var c in candidates)
            if (!string.IsNullOrEmpty(c) && File.Exists(c))
                return c;
        return null;
    }

    public static async Task<Result<bool>> InstallBepInEx(string gameRoot, Action<InstallProgress>? onProgress = null)
    {
        var archive = ResolveBepInExArchive();
        if (archive == null)
            return Result<bool>.Fail("BepInEx package is missing from the application bundle.");
        var tempRoot = ModArchiveExtractor.CreateTempDir(gameRoot);
        try
        {
            onProgress?.Invoke(new InstallProgress("setup-bepinex", 10, "Extracting BepInEx..."));
            await ModArchiveExtractor.ExtractArchiveAsync(archive, tempRoot);
            onProgress?.Invoke(new InstallProgress("setup-bepinex", 60, "Moving into game folder..."));
            MoveToGameRoot(tempRoot, Path.GetFullPath(gameRoot));
            onProgress?.Invoke(new InstallProgress("setup-bepinex", 100, "BepInEx installed."));
            return Result<bool>.Success(DetectBepInEx(gameRoot));
        }
        catch (Exception e)
        {
            return Result<bool>.Fail(e.Message);
        }
        finally
        {
            ModArchiveExtractor.RemoveDirIfInside(gameRoot, tempRoot);
        }
    }

    private static void MoveChildren(string srcDir, string dstDir)
    {
        foreach (var name in Directory.GetFileSystemEntries(srcDir))
        {
            var dst = Path.Combine(dstDir, Path.GetFileName(name));
            Directory.CreateDirectory(Path.GetDirectoryName(dst)!);
            try
            {
                if (Directory.Exists(name)) Directory.Move(name, dst);
                else File.Move(name, dst);
            }
            catch
            {
                if (Directory.Exists(name)) CopyDirectoryTree(name, dst);
                else File.Copy(name, dst, overwrite: true);
                try { if (Directory.Exists(name)) Directory.Delete(name, recursive: true); else File.Delete(name); }
                catch { /* best effort */ }
            }
        }
    }

    private static void MoveToGameRoot(string tempRoot, string gameRoot)
    {
        var top = Directory.GetFileSystemEntries(tempRoot);
        if (top.Length == 1 && Directory.Exists(top[0]) && Path.GetFileName(top[0]) != Constants.BepInExFolder)
        {
            MoveChildren(top[0], gameRoot);
            return;
        }
        MoveChildren(tempRoot, gameRoot);
    }

    private static void CopyDirectoryTree(string src, string dst)
    {
        Directory.CreateDirectory(dst);
        foreach (var entry in Directory.GetFileSystemEntries(src))
        {
            var to = Path.Combine(dst, Path.GetFileName(entry));
            if (Directory.Exists(entry)) CopyDirectoryTree(entry, to);
            else File.Copy(entry, to, overwrite: true);
        }
    }
}
